### PR TITLE
[SPARK-13263] [SQL] SQL Generation Support for Tablesample

### PR DESCRIPTION
**Pull request by gatorsmile**
In the parser, tableSample clause is part of tableSource. 

```
tableSource
@init { gParent.pushMsg("table source", state); }
@after { gParent.popMsg(state); }
    : tabname=tableName 
    ((tableProperties) => props=tableProperties)?
    ((tableSample) => ts=tableSample)? 
    ((KW_AS) => (KW_AS alias=Identifier) 
    |
    (Identifier) => (alias=Identifier))?
    -> ^(TOK_TABREF $tabname $props? $ts? $alias?)
    ;
```

Two typical query samples using TABLESAMPLE are:

```
    "SELECT s.id FROM t0 TABLESAMPLE(10 PERCENT) s"
    "SELECT * FROM t0 TABLESAMPLE(0.1 PERCENT)"
```

FYI, the logical plan of a TABLESAMPLE query: 

```
sql("SELECT * FROM t0 TABLESAMPLE(0.1 PERCENT)").explain(true)

== Analyzed Logical Plan ==
id: bigint
Project [id#16L]
+- Sample 0.0, 0.001, false, 381
   +- Subquery t0
      +- Relation[id#16L] ParquetRelation
```

Thanks! cc @liancheng

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala [hunk 1]
```diff
@@ -119,6 +120,18 @@ class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Loggi
         limitSQL = limit.sql
       } yield s"$childSQL LIMIT $limitSQL"
 
+    // TABLESAMPLE is part of tableSource clause in the parser,
+    // and thus we must handle it with subquery.
+    case Sample(lb, ub, withReplacement, _, child @ Subquery(alias, grandChild))
```
Comment (rxin) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: this seems pretty brittle. i wonder if we can change sample to make it very clear this is either a table sample, or a normal sample.

Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: Agree with @rxin. Especially, I don't think `MetastoreRelation` is wrapped by a `Subquery` in this case.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala [hunk 2]
```diff
@@ -145,6 +145,16 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
     checkHiveQl("SELECT COUNT(DISTINCT id) FROM t0")
   }
 
+  test("TABLESAMPLE") {
+    checkHiveQl("SELECT * FROM t0 TABLESAMPLE(100 PERCENT) s")
```
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: Testing tables created in this test suite are all Parquet tables, which are by default converted from `MetastoreRelation` to `ParquetRelation` wrapped with a `Subquery`. We should probably rename such tables, e.g., `t0` to `parquet_t0`, and add real text format `MetastoreRelation` testing tables via explicit HiveQL DDLs.

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: Thanks! Will do it.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: Thank you! Will follow your suggestions!

**Comment by gatorsmile** [comment]
@rxin After searching the code base, all the samples are table sample. Could you give me more hints? I also agree it looks brittle. I did my best to make it more robust/general. 

Since `TABLESAMPLE` is part of `tableSource` clause, I am facing two options: 

1. Simplify the logics in the case of `TABLESAMPLE` and put the extra handling in `MetastoreRelation` and `Subquery`
2. Put all the logics in the case of `TABLESAMPLE`.

Let me know which option you like. Currently, I picked option 2.

**Comment by rxin** [comment]
You can for example in the Sample operator add a method that returns true if it is created from table sample in the parser.

**Comment by rxin** [comment]
Basically it'd be better to have something very explicit, rather than relying on some conditions that might change in the future.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala [hunk 3]
```diff
@@ -80,6 +81,26 @@ class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Loggi
...
+          val aliasName = m.alias.getOrElse("")
+          build(
+            s"`${m.databaseName}`.`${m.tableName}`",
+            "TABLESAMPLE(" + fraction + " PERCENT)",
+            aliasName)
+        case s: Subquery =>
+          val aliasName = if (s.child.isInstanceOf[Subquery]) s.alias else ""
+          val plan = if (s.child.isInstanceOf[Subquery]) s.child else s
+          build(toSQL(plan), "TABLESAMPLE(" + fraction + " PERCENT)", aliasName)
+        case _ =>
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala]: Unable to create a test case for this branch. I do not know if we should throw an exception here or building a SQL.

**Comment by gatorsmile** [comment]
@rxin I see. Will do it. Thanks!

**Comment by rxin** [comment]
Looks more robust this way. I will let @liancheng do the final reviews and merge.

**Comment by gatorsmile** [comment]
Thanks for your reviews! @rxin

**Comment by gatorsmile** [comment]
ok to test

**Comment by liancheng** [comment]
LGTM, thanks!

Merged to master. There was a minor conflict. Resolved it manually while merging.

**Comment by gatorsmile** [comment]
Thank you! @liancheng @rxin

**Pull request closed without merging**